### DISCUSSION TITLE
If state has multiple actions, can the second action use first action output data?

### DISCUSSION THREAD

**Discussion by elim19** [serverlessworkflow/specification#752] [Q&A] [answered]
In https://github.com/serverlessworkflow/specification/blob/main/specification.md#using-multiple-data-filters, I am not seeing such examples

**Comment by tsurdilo** [comment, marked as answer]
Hi, are you asking for an example of this to be added to https://github.com/serverlessworkflow/specification/tree/main/examples ?

**Reply by elim19** [reply]
> Hi, are you asking for an example of this to be added to https://github.com/serverlessworkflow/specification/tree/main/examples ?

@tsurdilo, yes, but also I am not seeing how to pass data from first action to second in your https://github.com/serverlessworkflow/specification/blob/main/specification.md, is it possible?
because I am only seeing pass action data to state data in https://github.com/serverlessworkflow/specification/blob/main/specification.md#data-merging

**Reply by lsytj0413** [reply]
> > Hi, are you asking for an example of this to be added to https://github.com/serverlessworkflow/specification/tree/main/examples ?
> 
> @tsurdilo, yes, but also I am not seeing how to pass data from first action to second in your https://github.com/serverlessworkflow/specification/blob/main/specification.md, is it possible? because I am only seeing pass action data to state data in https://github.com/serverlessworkflow/specification/blob/main/specification.md#data-merging

I think it's not possible to directly pass data from first action to second. We should discuss the behavior on state's type:

1. Callback State
  - It only have one action to perform, so there is no question about this.
2. Event State
  - There is an `actionMode` which specifies how actions are to be performed (in sequence or in parallel), so it's parallel we can't directly pass data from first action to second.
3. ForEach State
  - have multi action to perform and no `actionMode` property.
4. Operation State
  - Have on `actionMode` property.
5. Parallel State
  - have multi action to perform and no `actionMode` property.

As we can see, if the State have multi action and with `actionMode` property, we can't directly pass data from first action to second. (because when it performed as parallel, the second action would run before first completed)；if the State didn't have `actionMode` property, swf specification didn't define what should be done.

**Reply by lsytj0413** [reply]
In my opinion，if you want to pass data from first action to second, we should do:

- set `actionMode` property to `sequence `
- each action write the output to stateData, use `useResults` & `toStateData` property.
- the follow action could use the result write by previous action

**Reply by elim19** [reply]
> In my opinion，if you want to pass data from first action to second, we should do:
> 
> * set `actionMode` property to `sequence `
> * each action write the output to stateData, use `useResults` & `toStateData` property.
> * the follow action could use the result write by previous action

@lsytj0413 , it is a workaround to use state as temporary storage, is it reasonable? because we may not want others to see these temporary data in states, then we need to use filter to clear these data in the state

**Reply by lsytj0413** [reply]
> @lsytj0413 , it is a workaround to use state as temporary storage, is it reasonable? because we may not want others to see these temporary data in states, then we need to use filter to clear these data in the state

I'm not clearly why you say `use state as temporary storage` and `need to use filter to clear these data in the state`, in one WorkflowInstance, we need StateData to pass data between each state and action, so it will persistent in runtime implement.

**Reply by elim19** [reply]
> > @lsytj0413 , it is a workaround to use state as temporary storage, is it reasonable? because we may not want others to see these temporary data in states, then we need to use filter to clear these data in the state
> 
> I'm not clearly why you say `use state as temporary storage` and `need to use filter to clear these data in the state`, in one WorkflowInstance, we need StateData to pass data between each state and action, so it will persistent in runtime implement.

@lsytj0413 , I am saying that because some data are stored in state just to pass to another action, not to another state, for example,
stateA has two key: actionTmpData and stateTmpData.
actionTmpData is the data to pass to other actions, stateTmpData is the data to pass to other states.
So when stateA transit to stateB, only stateTmpData is necessary, actionTmpData is not longer used

**Reply by lsytj0413** [reply]
> @lsytj0413 , I am saying that because some data are stored in state just to pass to another action, not to another state, for example, stateA has two key: actionTmpData and stateTmpData. actionTmpData is the data to pass to other actions, stateTmpData is the data to pass to other states. So when stateA transit to stateB, only stateTmpData is necessary, actionTmpData is not longer used

As this moment you can use stateDataFilter.output to discard actionTmpData before transit to stateB.

**Reply by tsurdilo** [reply]
yes as @lsytj0413 said :) 
action data filter has also useResults flag: https://github.com/serverlessworkflow/specification/blob/main/specification.md#action-data-filters
that could be used if you want to disregard the results of an action
we could improve bit on this front..will think about it
